### PR TITLE
[Core] Default shipping method basing on category fix

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | yes
| Related tickets | fixes https://github.com/Sylius/Sylius/pull/9388, related with https://github.com/Sylius/Sylius/pull/8629
| License         | MIT

Following changes fixes part of the problem described by @stefandoorn in https://github.com/Sylius/Sylius/pull/9388. As far as I see, when this PR will be merged, [one more scenario](https://github.com/Sylius/Sylius/pull/9388/files#diff-24783680797085d6500a51a2b5973548R35) would be solved by https://github.com/Sylius/Sylius/pull/8629

### CODE REVIEW

**Comment by stefandoorn** [comment]
Will this PR also land in 1.0.x? The issue occurs there. I'm (have to look into it again why) having issues upgrading to 1.1 (I think due to Lakion CMS plugin).

**Comment by stefandoorn** [comment]
I'm willing to test this PR @Zales0123, but as my project is on 1.0.x, it's rather hard as this one includes 1.1 & 1.2 upcoming changes.

**Comment by Zales0123** [comment]
Sorry for waiting @stefandoorn, there is a one problem with this PR being merged to `1.0` - it contains some deprecations which afaik cannot be avoided :/ I'm waiting for @pamil opinion about that and related PR, he should have some time in the nearest future to think about it (I hope so 😄)

**Comment by stefandoorn** [comment]
@pamil @Zales0123 Will this one make it into 1.0.x? I have some issues upgrading to 1.1/1.2 right now :(

**Comment by stefandoorn** [comment]
If it won't make it to 1.0.x btw, should we open a section in the docs about 'known issues'? This is a rather annoying bug for systems using multiple shipment methods / categories etc. The chances are high the wrong price is shown in the cart then.

**Comment by pamil** [comment]
@stefandoorn it should make it into v1.2.0 release (as it contains some deprecations I don't feel comfortable merging it into v1.0.x).

@Zales0123 could you rebase this PR?

**Comment by Zales0123** [comment]
@pamil @stefandoorn yes, sorry for the delay, I'm on it! 🏃

**Comment by stefandoorn** [comment]
Any updates on this one @Zales0123? If it's going to make it in 1.2.0, I will try to upgrade to 1.1.x & then to 1.2.0-beta today on this project :-)

**Comment by pamil** [comment]
@stefandoorn I've just rebased it, if it passes, it'll be included in v1.2.0.

File: src/Sylius/Component/Core/Resolver/CategoryBasedDefaultShippingMethodResolver.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+     */
+    public function __construct(
+        ShippingMethodRepositoryInterface $shippingMethodRepository,
+        ShippingMethodEligibilityCheckerInterface $shippingMethodEligibilityChecker,
+        ?ZoneMatcherInterface $zoneMatcher = null
+    ) {
+        $this->shippingMethodRepository = $shippingMethodRepository;
+        $this->shippingMethodEligibilityChecker = $shippingMethodEligibilityChecker;
+
+        if (2 === func_num_args() || null === $zoneMatcher) {
```
Comment (pamil) [inline src/Sylius/Component/Core/Resolver/CategoryBasedDefaultShippingMethodResolver.php]: It looks like this class is brand new so we don't need to introduce a deprecation here (and we should require that zone matcher).

**Comment by Zales0123** [comment]
@pamil @stefandoorn it should work like a charm now, again, sorry for the delay :)

File: UPGRADE-1.2.md [hunk 2]
```diff
@@ -9,3 +9,5 @@
 * The deprecated form mapping feature in SonataCoreBundle has been disabled in the app configuration included from SyliusCoreBundle.
   If you depend on the feature in your application, you will need to make the necessary changes. Refer to
   https://github.com/sonata-project/SonataCoreBundle/pull/462 for more information. 
+
+* Service with id `sylius.shipping_method_resolver.default` has been deprecated and will be removed in 2.0. `sylius.shipping_method_resolver.category_based` should be used instead.
```
[the remark this answers is not in the document]
Comment (Zales0123) [reply UPGRADE-1.2.md]: What solution do you suggest?

[the remark this answers is not in the document]
Comment (pamil) [reply UPGRADE-1.2.md]: Registering that `CategoryBased...` thingie as `sylius.shipping_method_resolver.default` service would fix that (and removing `sylius.shipping_method_resolver.category_based` service).

[the remark this answers is not in the document]
Comment (pamil) [reply UPGRADE-1.2.md]: Going further, we can deprecate CoreBundle's `DefaultShippingMethodResolver` (adding a `@trigger_error(..., E_USER_DEPRECATED)` before the class declaration, we need to look at Symfony for the exact code) and remove the deprecation/changes made earlier there in https://github.com/Sylius/Sylius/pull/8629.

File: src/Sylius/Component/Core/Resolver/CategoryBasedDefaultShippingMethodResolver.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * @param ShippingMethodRepositoryInterface $shippingMethodRepository
+     * @param ShippingMethodEligibilityCheckerInterface $shippingMethodEligibilityChecker
+     * @param ZoneMatcherInterface $zoneMatcher
+     */
+    public function __construct(
+        ShippingMethodRepositoryInterface $shippingMethodRepository,
+        ShippingMethodEligibilityCheckerInterface $shippingMethodEligibilityChecker,
+        ?ZoneMatcherInterface $zoneMatcher
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Resolver/CategoryBasedDefaultShippingMethodResolver.php]: Should it be an optional parameter? Either we need to remove question mark and maybe change the name to `CategoryAndZoneBasedDefaultShippingMethodResolver` or change `ZoneMatcherInterface` typehint to `ZoneMatcherInterface|null` and support nullable value in line 97.

Comment (lchrusciel) [reply src/Sylius/Component/Core/Resolver/CategoryBasedDefaultShippingMethodResolver.php]: And change the name to `CategoryAndPerhapsZoneBasedDefaultShippingMethodResolver` 🐋

File: src/Sylius/Component/Core/Resolver/DefaultShippingMethodResolver.php [hunk 4]
```diff
@@ -33,27 +35,20 @@ class DefaultShippingMethodResolver implements DefaultShippingMethodResolverInte
...
     /**
-     * @var ZoneMatcherInterface|null
+     * @var ZoneMatcherInterface
      */
     private $zoneMatcher;
 
     /**
      * @param ShippingMethodRepositoryInterface $shippingMethodRepository
-     * @param ZoneMatcherInterface|null $zoneMatcher
+     * @param ZoneMatcherInterface $zoneMatcher
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Resolver/DefaultShippingMethodResolver.php]: Shouldn't we leave this class as it is (even if buggy)?

File: src/Sylius/Component/Core/Resolver/CategoryBasedDefaultShippingMethodResolver.php (hunk 3 above)
Comment (Zales0123) [reply src/Sylius/Component/Core/Resolver/CategoryBasedDefaultShippingMethodResolver.php]: It should not be optional. Will be fixed in a minute

File: src/Sylius/Component/Core/OrderProcessing/OrderShipmentProcessor.php [hunk 5]
```diff
@@ -76,48 +76,59 @@ public function process(BaseOrderInterface $order): void
...
-        }
-
         try {
             /** @var ShipmentInterface $shipment */
             $shipment = $this->shipmentFactory->createNew();
             $shipment->setOrder($order);
+
+            $this->processShipmentUnits($order, $shipment);
+
             $shipment->setMethod($this->defaultShippingMethodResolver->getDefaultShippingMethod($shipment));
```
[the remark this answers is not in the document]
Comment (Zales0123) [reply src/Sylius/Component/Core/OrderProcessing/OrderShipmentProcessor.php]: I'd leave it as it is right now (it works fine) and refactor it in next PR (this line is not even changed of this PR)

File: src/Sylius/Component/Core/Resolver/CategoryBasedDefaultShippingMethodResolver.php (hunk 3 above)
Comment (pamil) [reply src/Sylius/Component/Core/Resolver/CategoryBasedDefaultShippingMethodResolver.php]: `EligibleDefaultShippingMethodResolver`? And yeah, zone matcher should be required.

File: src/Sylius/Component/Core/Resolver/DefaultShippingMethodResolver.php (hunk 4 above)
Comment (Zales0123) [reply src/Sylius/Component/Core/Resolver/DefaultShippingMethodResolver.php]: nope, see [this comment](https://github.com/Sylius/Sylius/pull/9398#discussion_r193361059) 😄

File: src/Sylius/Component/Core/Resolver/DefaultShippingMethodResolver.php [hunk 6]
```diff
@@ -25,6 +25,8 @@
 use Sylius\Component\Shipping\Resolver\DefaultShippingMethodResolverInterface;
 use Webmozart\Assert\Assert;
 
+@trigger_error('This class is deprecated since Sylius 1.2 and will be removed in 2.0. "Sylius\Component\Core\Resolver\CategoryBasedDefaultShippingMethodResolver" should be used instead.', E_USER_DEPRECATED);
```
Comment (pamil) [inline src/Sylius/Component/Core/Resolver/DefaultShippingMethodResolver.php]: `"Sylius\Component\Core\Resolver\CategoryBasedDefaultShippingMethodResolver"` screams for `sprintf("%s", CategoryBasedDefaultShippingMethodResolver::class)`

File: features/cart/shopping_cart/resolving_default_shipping_method_in_cart_summary_for_simple_products.feature [hunk 7]
```diff
@@ -0,0 +1,35 @@
+@shopping_cart
+Feature: Viewing a cart summary with correct default shipping method based on variant applied
+    In order to see details about my order
+    As a Visitor
+    I want to be able to see my cart summary with the correct shipping method based on variants
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has "Over-sized" shipping category
```
[the remark this answers is not in the document]
Comment (Zales0123) [reply features/cart/shopping_cart/resolving_default_shipping_method_in_cart_summary_for_simple_products.feature]: Yes, it's used in the shipping method below.

File: src/Sylius/Component/Core/Resolver/DefaultShippingMethodResolver.php [hunk 8]
```diff
@@ -33,27 +35,20 @@ class DefaultShippingMethodResolver implements DefaultShippingMethodResolverInte
...
 
     /**
      * @param ShippingMethodRepositoryInterface $shippingMethodRepository
-     * @param ZoneMatcherInterface|null $zoneMatcher
+     * @param ZoneMatcherInterface $zoneMatcher
      */
     public function __construct(
         ShippingMethodRepositoryInterface $shippingMethodRepository,
-        ?ZoneMatcherInterface $zoneMatcher = null
+        ZoneMatcherInterface $zoneMatcher
```
Comment (pamil) [inline src/Sylius/Component/Core/Resolver/DefaultShippingMethodResolver.php]: We should remove this argument (revert #8629 changes), the current version is a BC break :)

**Comment by Zales0123** [comment]
ping @pamil @lchrusciel @stefandoorn

File: UPGRADE-1.2.md [hunk 9]
```diff
@@ -9,3 +9,5 @@
 * The deprecated form mapping feature in SonataCoreBundle has been disabled in the app configuration included from SyliusCoreBundle.
   If you depend on the feature in your application, you will need to make the necessary changes. Refer to
   https://github.com/sonata-project/SonataCoreBundle/pull/462 for more information. 
+
+* Class `Sylius\Component\Core\Resolver\DefaultShippingMethodResolver` has been deprecated and will be removed in 2.0. `Sylius\Component\Core\Resolver\CategoryBasedDefaultShippingMethodResolver` should be used instead.
```
Comment (pamil) [inline UPGRADE-1.2.md]: `CategoryBased...` -> `EligibleDefaultShippingMethodResolver`

Comment (Zales0123) [reply UPGRADE-1.2.md]: aaaaaa :(

Comment (Zales0123) [reply UPGRADE-1.2.md]: @pamil fixed

**Comment by pamil** [comment]
Thank you, Matthias! 🥇

**Pull request merged**